File: tests/algorithms/x/Pascal/backtracking/power_sum.pas
{$mode objfpc}
program Main;
uses SysUtils;
var _nowSeed: int64 = 0;
var _nowSeeded: boolean = false;
procedure init_now();
var s: string; v: int64;
begin
  s := GetEnvironmentVariable('MOCHI_NOW_SEED');
  if s <> '' then begin
    Val(s, v);
    _nowSeed := v;
    _nowSeeded := true;
  end;
end;
function _now(): integer;
begin
  if _nowSeeded then begin
    _nowSeed := (_nowSeed * 1664525 + 1013904223) mod 2147483647;
    _now := _nowSeed;
  end else begin
    _now := Integer(GetTickCount64()*1000);
  end;
end;
function _bench_now(): int64;
begin
  _bench_now := GetTickCount64()*1000;
end;
function _mem(): int64;
var h: TFPCHeapStatus;
begin
  h := GetFPCHeapStatus;
  _mem := h.CurrHeapUsed;
end;
var
  bench_start_0: integer;
  bench_dur_0: integer;
  bench_mem_0: int64;
  bench_memdiff_0: int64;
function int_pow(base: integer; exp: integer): integer; forward;
function backtrack(target: integer; exp: integer; current: integer; current_sum: integer): integer; forward;
function solve(target: integer; exp: integer): integer; forward;
function int_pow(base: integer; exp: integer): integer;
var
  int_pow_result_: integer;
  int_pow_i: integer;
begin
  int_pow_result_ := 1;
  int_pow_i := 0;
  while int_pow_i < exp do begin
  int_pow_result_ := int_pow_result_ * base;
  int_pow_i := int_pow_i + 1;
end;
  exit(int_pow_result_);
end;
function backtrack(target: integer; exp: integer; current: integer; current_sum: integer): integer;
var
  backtrack_p: integer;
  backtrack_count: integer;
begin
  if current_sum = target then begin
  exit(1);
end;
  backtrack_p := int_pow(current, exp);
  backtrack_count := 0;
  if (current_sum + backtrack_p) <= target then begin
  backtrack_count := backtrack_count + backtrack(target, exp, current + 1, current_sum + backtrack_p);
end;
  if backtrack_p < target then begin
  backtrack_count := backtrack_count + backtrack(target, exp, current + 1, current_sum);
end;
  exit(backtrack_count);
end;
function solve(target: integer; exp: integer): integer;
begin
  if not ((((1 <= target) and (target <= 1000)) and (2 <= exp)) and (exp <= 10)) then begin
  writeln('Invalid input');
  exit(0);
end;
  exit(backtrack(target, exp, 1, 0));
end;
begin
  init_now();
  bench_mem_0 := _mem();
  bench_start_0 := _bench_now();
  writeln(solve(13, 2));
  writeln(solve(10, 2));
  writeln(solve(10, 3));
  Sleep(1);
  bench_memdiff_0 := _mem() - bench_mem_0;
  bench_dur_0 := (_bench_now() - bench_start_0) div 1000;
  writeln('{');
  writeln(('  "duration_us": ' + IntToStr(bench_dur_0)) + ',');
  writeln(('  "memory_bytes": ' + IntToStr(bench_memdiff_0)) + ',');
  writeln(('  "name": "' + 'main') + '"');
  writeln('}');
end.

File: tests/algorithms/transpiler/FS/backtracking/power_sum.fs
// Generated 2025-08-06 20:48 +0700

exception Return
let mutable _nowSeed:int64 = 0L
let mutable _nowSeeded = false
let _initNow () =
    let s = System.Environment.GetEnvironmentVariable("MOCHI_NOW_SEED")
    if System.String.IsNullOrEmpty(s) |> not then
        match System.Int32.TryParse(s) with
        | true, v ->
            _nowSeed <- int64 v
            _nowSeeded <- true
        | _ -> ()
let _now () =
    if _nowSeeded then
        _nowSeed <- (_nowSeed * 1664525L + 1013904223L) % 2147483647L
        int _nowSeed
    else
        int (System.DateTime.UtcNow.Ticks % 2147483647L)

_initNow()
let __bench_start = _now()
let __mem_start = System.GC.GetTotalMemory(true)
let rec int_pow (``base``: int) (exp: int) =
    let mutable __ret : int = Unchecked.defaultof<int>
    let mutable ``base`` = ``base``
    let mutable exp = exp
    try
        let mutable result: int = 1
        let mutable i: int = 0
        while i < exp do
            result <- result * ``base``
            i <- i + 1
        __ret <- result
        raise Return
        __ret
    with
        | Return -> __ret
and backtrack (target: int) (exp: int) (current: int) (current_sum: int) =
    let mutable __ret : int = Unchecked.defaultof<int>
    let mutable target = target
    let mutable exp = exp
    let mutable current = current
    let mutable current_sum = current_sum
    try
        if current_sum = target then
            __ret <- 1
            raise Return
        let p: int = int_pow (current) (exp)
        let mutable count: int = 0
        if (current_sum + p) <= target then
            count <- count + (backtrack (target) (exp) (current + 1) (current_sum + p))
        if p < target then
            count <- count + (backtrack (target) (exp) (current + 1) (current_sum))
        __ret <- count
        raise Return
        __ret
    with
        | Return -> __ret
and solve (target: int) (exp: int) =
    let mutable __ret : int = Unchecked.defaultof<int>
    let mutable target = target
    let mutable exp = exp
    try
        if not ((((1 <= target) && (target <= 1000)) && (2 <= exp)) && (exp <= 10)) then
            printfn "%s" ("Invalid input")
            __ret <- 0
            raise Return
        __ret <- backtrack (target) (exp) (1) (0)
        raise Return
        __ret
    with
        | Return -> __ret
printfn "%d" (solve (13) (2))
printfn "%d" (solve (10) (2))
printfn "%d" (solve (10) (3))
let __bench_end = _now()
let __mem_end = System.GC.GetTotalMemory(true)
printfn "{\n  \"duration_us\": %d,\n  \"memory_bytes\": %d,\n  \"name\": \"main\"\n}" ((__bench_end - __bench_start) / 1000) (__mem_end - __mem_start)
